<!doctype html>
<html>
<head>
<meta charset="UTF-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<meta http-equiv="X-UA-Compatible" content="ie=edge" />
<title>Markmap</title>
<style>
* {
  margin: 0;
  padding: 0;
}
html {
  font-family: ui-sans-serif, system-ui, sans-serif, 'Apple Color Emoji',
    'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
}
#mindmap {
  display: block;
  width: 100vw;
  height: 100vh;
}
.markmap-dark {
  background: #27272a;
  color: white;
}
</style>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/markmap-toolbar@0.18.12/dist/style.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@highlightjs/cdn-assets@11.11.1/styles/default.min.css">
</head>
<body>
<svg id="mindmap"></svg>
<script src="https://cdn.jsdelivr.net/npm/d3@7.9.0/dist/d3.min.js"></script><script src="https://cdn.jsdelivr.net/npm/markmap-view@0.18.12/dist/browser/index.js"></script><script src="https://cdn.jsdelivr.net/npm/markmap-toolbar@0.18.12/dist/index.js"></script><script>(r => {
              setTimeout(r);
            })(function renderToolbar() {
  const {
    markmap,
    mm
  } = window;
  const {
    el
  } = markmap.Toolbar.create(mm);
  el.setAttribute('style', 'position:absolute;bottom:20px;right:20px');
  document.body.append(el);
})</script><script>((getMarkmap, getOptions, root2, jsonOptions) => {
              const markmap = getMarkmap();
              window.mm = markmap.Markmap.create(
                "svg#mindmap",
                (getOptions || markmap.deriveOptions)(jsonOptions),
                root2
              );
              if (window.matchMedia("(prefers-color-scheme: dark)").matches) {
                document.documentElement.classList.add("markmap-dark");
              }
            })(() => window.markmap,null,{"content":"","children":[{"content":"<strong>Using try and catch Statements</strong>","children":[{"content":"Java uses a <code>try</code> statement to separate the logic that<br>\nmight throw an exception from the logic to handle that<br>\nexception.","children":[{"content":"The syntax of a <code>try</code> statement","children":[{"content":"\n<pre data-lines=\"11,20\"><code class=\"language-js\"><span class=\"hljs-keyword\">try</span> {<span class=\"hljs-comment\">//try keyword and open and close braces are required</span>\n    <span class=\"hljs-comment\">// Protected code</span>\n} <span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">Exception</span> e){    <span class=\"hljs-comment\">//catch keyword</span>\n                          <span class=\"hljs-comment\">// The type of exception being caught</span>\n                          <span class=\"hljs-comment\">//The identifier of the exception object</span>\n    <span class=\"hljs-comment\">// Exception handler </span>\n}\n</code></pre>","children":[],"payload":{"tag":"li","lines":"11,20"}},{"content":"The code in the <code>try</code> block is run normally. If any of the statements throws an exception<br>\nthat can be caught by the exception type listed in the <code>catch</code> block, the <code>try</code> block stops<br>\nrunning, and execution goes to the <code>catch</code> statement. If none of the statements in the<br>\n<code>try</code> block throws an exception that can be caught, the <code>catch</code> clause is not run.","children":[],"payload":{"tag":"li","lines":"20,24"}}],"payload":{"tag":"li","lines":"10,24"}}],"payload":{"tag":"li","lines":"7,24"}},{"content":"\n<pre data-lines=\"24,36\"><code class=\"language-js\"><span class=\"hljs-number\">3</span>: <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">explore</span>(<span class=\"hljs-params\"></span>) {\n<span class=\"hljs-number\">4</span>:   <span class=\"hljs-keyword\">try</span> {\n<span class=\"hljs-number\">5</span>:     <span class=\"hljs-title function_\">fall</span>();\n<span class=\"hljs-number\">6</span>:     <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(<span class=\"hljs-string\">&quot;never get here&quot;</span>);\n<span class=\"hljs-number\">7</span>:   } <span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">RuntimeException</span> e) {\n<span class=\"hljs-number\">8</span>:     <span class=\"hljs-title function_\">getUp</span>();\n<span class=\"hljs-number\">9</span>:  }\n<span class=\"hljs-number\">10</span>:   <span class=\"hljs-title function_\">seeAnimals</span>();\n<span class=\"hljs-number\">11</span>: }\n<span class=\"hljs-number\">12</span>: <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">fall</span>(<span class=\"hljs-params\"></span>) { <span class=\"hljs-keyword\">throw</span> <span class=\"hljs-keyword\">new</span> <span class=\"hljs-title class_\">RuntimeException</span>(); }\n</code></pre>","children":[{"content":"The curly braces are required<br>\nfor <code>try</code> and <code>catch</code> blocks.","children":[{"content":"\n<pre data-lines=\"38,45\"><code data-lines=\"38,45\"><span class=\"hljs-keyword\">try</span>\n<span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n  fall();\n<span class=\"hljs-keyword\">catch</span> (Exception e)\n  System.<span class=\"hljs-keyword\">out</span>.println(<span class=\"hljs-string\">&quot;get up&quot;</span>);\n</code></pre>","children":[],"payload":{"tag":"li","lines":"38,45"}},{"content":"\n<pre data-lines=\"45,50\"><code data-lines=\"45,50\"><span class=\"hljs-keyword\">try</span> { <span class=\"hljs-regexp\">// DOES NOT COMPILE\n  fall();\n} //</span><span class=\"hljs-keyword\">catch</span> o &#xfb01;nally <span class=\"hljs-keyword\">is</span> missing\n</code></pre>","children":[],"payload":{"tag":"li","lines":"45,50"}}],"payload":{"tag":"li","lines":"36,50"}}],"payload":{"tag":"li","lines":"24,50"}}],"payload":{"tag":"h1","lines":"6,7"}},{"content":"<strong>Chaining catch Blocks</strong>","children":[{"content":"\n<p data-lines=\"51,54\">First, you must be able to recognize if the exception is a checked<br>\nor an unchecked exception. Second, you need to determine<br>\nwhether any of the exceptions are subclasses of the others.</p>","children":[],"payload":{"tag":"li","lines":"51,54"}},{"content":"\n<p data-lines=\"54,57\"><code>catch</code> blocks from top to bottom. First place the child exceptions<br>\nand then the parent exception; if there is no inheritance relationship<br>\nthen they can go in any order.</p>","children":[{"content":"\n<pre data-lines=\"57,77\"><code class=\"language-js\"><span class=\"hljs-keyword\">class</span> <span class=\"hljs-title class_\">AnimalsOutForAWalk</span> <span class=\"hljs-keyword\">extends</span> <span class=\"hljs-title class_ inherited__\">RuntimeException</span> {}\n<span class=\"hljs-keyword\">class</span> <span class=\"hljs-title class_\">ExhibitClosed</span> <span class=\"hljs-keyword\">extends</span> <span class=\"hljs-title class_ inherited__\">RuntimeException</span> {}\n<span class=\"hljs-keyword\">class</span> <span class=\"hljs-title class_\">ExhibitClosedForLunch</span> <span class=\"hljs-keyword\">extends</span> <span class=\"hljs-title class_ inherited__\">ExhibitClosed</span> {}\npublic <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">test</span>(<span class=\"hljs-params\"></span>) {\n  <span class=\"hljs-keyword\">try</span> {\n        <span class=\"hljs-title function_\">seeAnimal</span>();\n  }<span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">AnimalsOutForAWalk</span> e){ <span class=\"hljs-comment\">// first catch block</span>\n        <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">print</span>(<span class=\"hljs-string\">&quot;try back later&quot;</span>);\n  }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">ExhibitClosedForLunch</span> e){ <span class=\"hljs-comment\">// second catch block</span>\n        <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(<span class=\"hljs-string\">&quot;having lunch&quot;</span>);\n  }<span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">ExhibitClosed</span> e){ <span class=\"hljs-comment\">// third catch block</span>\n        <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">print</span>(<span class=\"hljs-string\">&quot;not today&quot;</span>);\n  }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">RuntimeException</span> e){ <span class=\"hljs-comment\">// fourth catch block</span>\n        <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(<span class=\"hljs-string\">&quot;RuntimeException&quot;</span>);\n  }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">Exception</span> e){ <span class=\"hljs-comment\">//fifth catch block</span>\n        <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(<span class=\"hljs-string\">&quot;Exception&quot;</span>);\n      }\n}\n</code></pre>","children":[{"content":"The order of the first and second <code>catch</code> blocks could be<br>\nreversed because the exceptions don&#x2019;t inherit from each<br>\nother.  For the others, there is an inheritance relationship.","children":[],"payload":{"tag":"li","lines":"77,80"}},{"content":"A rule exists for the order of the <code>catch</code> blocks. Java looks<br>\nat them in the order they appear. If it is impossible for one<br>\nof the <code>catch</code> blocks to be executed, a compiler error about<br>\nunreachable code occurs.","children":[],"payload":{"tag":"li","lines":"80,84"}},{"content":"It is not possible to execute all the catch blocks, only one,<br>\nif it is compatible with the exception thrown.","children":[],"payload":{"tag":"li","lines":"84,86"}}],"payload":{"tag":"li","lines":"57,86"}}],"payload":{"tag":"li","lines":"54,86"}},{"content":"\n<p data-lines=\"86,87\">The following example shows exception types that do inherit from each other:</p>\n<pre data-lines=\"88,99\"><code class=\"language-js\">public <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">visitMonkeys</span>(<span class=\"hljs-params\"></span>) {\n    <span class=\"hljs-keyword\">try</span> {\n      <span class=\"hljs-title function_\">seeAnimal</span>();\n    }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">ExhibitClosedForLunch</span> e){ <span class=\"hljs-comment\">// Subclass exception</span>\n      <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">print</span>(<span class=\"hljs-string\">&quot;try back later&quot;</span>);\n    }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">ExhibitClosed</span> e){        <span class=\"hljs-comment\">// Superclass exception</span>\n      <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">print</span>(<span class=\"hljs-string\">&quot;not today&quot;</span>);\n  }\n}\n</code></pre>","children":[{"content":"The reverse does not work","children":[{"content":"\n<pre data-lines=\"100,111\"><code data-lines=\"100,111\"><span class=\"hljs-function\"><span class=\"hljs-keyword\">public</span> <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title\">visitMonkeys</span>()</span> {\n  <span class=\"hljs-keyword\">try</span> {\n    seeAnimal();\n  } <span class=\"hljs-keyword\">catch</span> (ExhibitClosed e){\n    System.<span class=\"hljs-keyword\">out</span>.print(<span class=\"hljs-string\">&quot;not today&quot;</span>);\n  }<span class=\"hljs-keyword\">catch</span>(ExhibitClosedForLunch e){ <span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n      System.<span class=\"hljs-keyword\">out</span>.print(<span class=\"hljs-string\">&quot;try back later&quot;</span>);\n  }\n}\n</code></pre>","children":[],"payload":{"tag":"li","lines":"100,111"}},{"content":"If the more specific <code>ExhibitClosedForLunch</code> exception is<br>\nthrown, the <code>catch</code> block for <code>ExhibitClosed</code> runs&#x2014; &#xad;which<br>\nmeans there is no way for the second <code>catch</code> block to<br>\never run. Java correctly tells you there is an unreachable<br>\ncatch block.","children":[],"payload":{"tag":"li","lines":"111,116"}}],"payload":{"tag":"li","lines":"99,116"}}],"payload":{"tag":"li","lines":"86,116"}},{"content":"\n<pre data-lines=\"116,124\"><code class=\"language-js\">public <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">visit</span>(<span class=\"hljs-params\"></span>) {\n  <span class=\"hljs-keyword\">try</span> {\n    }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">IllegalArgumentException</span> e){\n    }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">NumberFormatException</span> e){ <span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n  }\n}\n</code></pre>","children":[{"content":"Second <code>catch</code> block unreachable","children":[],"payload":{"tag":"li","lines":"124,125"}},{"content":"Remember","children":[{"content":"<code>NumberFormatException</code> is<br>\na subclass of <code>IllegalArgumentException</code>","children":[],"payload":{"tag":"li","lines":"126,128"}},{"content":"<code>FileNotFoundException</code> is a subclass of <code>IOException</code>","children":[],"payload":{"tag":"li","lines":"128,129"}}],"payload":{"tag":"li","lines":"125,129"}}],"payload":{"tag":"li","lines":"116,129"}},{"content":"\n<p data-lines=\"129,131\">Remember that an exception defined by the <code>catch</code><br>\nstatement is only in scope for that <code>catch</code> block.</p>","children":[{"content":"\n<pre data-lines=\"131,141\"><code class=\"language-js\">public <span class=\"hljs-keyword\">void</span> <span class=\"hljs-title function_\">visit</span>(<span class=\"hljs-params\"></span>) {\n  <span class=\"hljs-keyword\">try</span> {\n  }<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">NumberFormatException</span> e1) {\n    <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(e1);\n  } <span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">IllegalArgumentException</span> e2) {\n    <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(e1); <span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n  }\n}\n</code></pre>","children":[],"payload":{"tag":"li","lines":"131,141"}}],"payload":{"tag":"li","lines":"129,141"}}],"payload":{"tag":"h1","lines":"50,51"}},{"content":"<strong>Applying a Multi-&#xad;catch Block</strong>","children":[{"content":"A <em>multiple catch block</em> allows multiple types of exceptions (with no<br>\ninheritance relationship) to be caught by the same catch block.","children":[],"payload":{"tag":"li","lines":"142,144"}},{"content":"\n<pre data-lines=\"144,152\"><code class=\"language-js\"><span class=\"hljs-title class_\">String</span> name =  <span class=\"hljs-literal\">null</span>;\n<span class=\"hljs-keyword\">try</span>{\n    name.<span class=\"hljs-title function_\">toString</span>();\n}<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">NullPointerException</span> | <span class=\"hljs-title class_\">IndexOutOfBoundsException</span> e){\n    <span class=\"hljs-title class_\">System</span>.<span class=\"hljs-property\">out</span>.<span class=\"hljs-title function_\">println</span>(e);\n}\n</code></pre>","children":[{"content":"If you wanted, you could still have a second <code>catch</code> block for <code>Exception</code><br>\nin case you want to handle other types of exceptions differently.","children":[],"payload":{"tag":"li","lines":"152,154"}},{"content":"It is like a regular <code>catch</code> clause, except two or more exception types are specified,<br>\nseparated by a pipe. The pipe (<code>|</code>) is also used as the &#x201c;or&#x201d; operator, making it easy<br>\nto remember that you can use either/or of the exception types. Notice how there is<br>\nonly one variable name in the catch clause. Java is saying that the variable named<br>\n<code>e</code> can be of type <code>NullPointerException</code> or <code>IndexOutOfBoundsException</code>.","children":[],"payload":{"tag":"li","lines":"154,159"}}],"payload":{"tag":"li","lines":"144,159"}},{"content":"Exceptions can be listed in any order within the <code>catch</code> clause. However,<br>\nthe variable name must appear only once and at the end.","children":[{"content":"\n<pre data-lines=\"161,166\"><code data-lines=\"161,166\"><span class=\"hljs-function\"><span class=\"hljs-title\">catch</span><span class=\"hljs-params\">(Exception1 e | Exception2 e | Exception3 e)</span></span> <span class=\"hljs-comment\">// DOES NOT COMPILE`</span>\n<span class=\"hljs-function\"><span class=\"hljs-title\">catch</span><span class=\"hljs-params\">(Exception1 e1 | Exception2 e2 | Exception3 e3)</span></span> <span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n<span class=\"hljs-function\"><span class=\"hljs-title\">catch</span><span class=\"hljs-params\">(Exception1 | Exception2 | Exception3 e)</span></span><span class=\"hljs-comment\">// CORRECT SYNTAX</span>\n</code></pre>","children":[],"payload":{"tag":"li","lines":"161,166"}}],"payload":{"tag":"li","lines":"159,166"}},{"content":"<em>Multi-&#xad;catch</em> to be used for exceptions that aren&apos;t related, and it<br>\nprevents you from specifying redundant types in a multi-&#xad;catch.","children":[{"content":"\n<pre data-lines=\"168,173\"><code class=\"language-js\"><span class=\"hljs-keyword\">try</span> {\n  <span class=\"hljs-keyword\">throw</span> <span class=\"hljs-keyword\">new</span> <span class=\"hljs-title class_\">IOException</span>();\n}<span class=\"hljs-keyword\">catch</span>(<span class=\"hljs-title class_\">FileNotFoundException</span> | <span class=\"hljs-title class_\">IOException</span> p) {} <span class=\"hljs-comment\">// DOES NOT COMPILE</span>\n</code></pre>","children":[{"content":"The exception <code>FileNotFoundException</code> is<br>\nalready caught by the alternative <code>IOException</code>","children":[],"payload":{"tag":"li","lines":"173,175"}},{"content":"Correct:<pre data-lines=\"176,181\"><code class=\"language-js\"><span class=\"hljs-keyword\">try</span> {\n  <span class=\"hljs-keyword\">throw</span> <span class=\"hljs-keyword\">new</span> <span class=\"hljs-title class_\">IOException</span>();\n} <span class=\"hljs-keyword\">catch</span> (<span class=\"hljs-title class_\">IOException</span> e) {}\n</code></pre>","children":[],"payload":{"tag":"li","lines":"175,181"}}],"payload":{"tag":"li","lines":"168,181"}}],"payload":{"tag":"li","lines":"166,181"}}],"payload":{"tag":"h1","lines":"141,142"}}]},{"initialExpandLevel":2})</script>
</body>
</html>
